proversity-org/edx-platform
# CAM-5 Adding new SAML backend

# CAM-5

## Description 
This PR make a suggestion for the username based on the full name in the registration form.

## How to test

1. Configure a SAML provider such as Testshib

Step by step:

* SAML does not work with localhost, first of all, you'll need to configure ngrok or localtunnel and will give you a URL, save it for later.

* Generate a private and public key in your local computer (`openssl req -new -x509 -days 3652 -nodes -out saml.crt -keyout saml.key`)

* Go to `/admin/third_party_auth/samlconfiguration/` and add a new one.

* Once inside do the following:
            **1.** Check enable box
            **2.** Add a new site with the URL provided by ngrok (or any that you use)
            **3.** Paste the values of the saml.key and saml.cert respectively in private and public key
            **4.** In `entity id` paste again the URL provided by ngrok but add the string saml at the 
                     beginning, for example `https://saml.9dd0e842.ngrok.io`
              **5.** In `Organization info` you'll see a JSON, in `url` key paste again the URL provided by 
                        ngrok without any modifications. Save it.
* Go to `/admin/third_party_auth/samlproviderconfig/` and create a new one. Once inside do the following:
**1.** Enable the checkbox
**2.** Give a name
**3.** Select the site that you create in the previous step
**4.** Check the `visible` checkbox
**5.** In `idp-slug` give it any string without whitespaces.
**6.** In `entity ID` paste exactly this: `https://idp.testshib.org/idp/shibboleth`
**7.** In `Metadata source` paste exactly this: `https://www.testshib.org/metadata/testshib-providers.xml`
**8.** In `Entity provider type` choose `Standard SAML provider`

* By last go to `https://your_url_in_ngrok/auth/saml/metadata.xml` save this XML in your local.
* Go to `https://www.testshib.org/register.html` and upload the XML that you just saved.
* Go to `https://your_url_in_ngrok/resgister` and you'll see the button and if all is ok, you'll be able to register without any problem. Since we are using Ngrok, please reference the 4 step with the "important" note.

2. Add the new backend (`third_party_auth.saml.HintedUsernameSAMLAuthBackend`) to the list of backends immediately **after** `'third_party_auth.saml.SAMLAuthBackend'`.

3. Optional, in settings or in site_configurations add the following settings:

    If in settings:
    `FEATURES["USERNAME_GENERATOR"] = {'SEPARATOR':'_', 'LOWER':True, 'RANDOM': False}`

    If in site_configurations:
    `{"USERNAME_GENERATOR" :  {'SEPARATOR':'_', 'LOWER':True, 'RANDOM': False}}`

    SEPARATOR key only receives a string. LOWER and RANDOM only receives a bool.

    This is optional because by default takes  `_` as separator character and LOWER and RANDOM as True

    For production, these settings have to be configured in `lms.env.json`

4. **Important:** If you going to test it with a proxy such as Ngrok you have to make a mandatory small modification. In `common/third_party_auth/strategy.py` add the following method to the class `ConfigurationModelStrategy`:
    ```
    def request_port(self):
        return 443
    ```

    This a bug in Python-SAML which returns 8000 port, we need to force it to return with SSL.

## Post-approval actions

## Reviewers
- [ ] @jagonzalr
